routers/api/v1/repo/action.go

```diff
@@ -581,3 +587,333 @@ func ListActionTasks(ctx *context.APIContext) {
+	if art.ArtifactName+".zip" == art.ArtifactPath && art.ContentEncoding == "application/zip" {
+		art := art
+		if setting.Actions.ArtifactStorage.ServeDirect() {
+			u, err := storage.ActionsArtifacts.URL(art.StoragePath, art.ArtifactPath, nil)
+			if u != nil && err == nil {
+				ctx.Redirect(u.String(), 302)
+				return
+			}
+		}
+		// ##[error]Unable to download artifact(s): Unable to download artifact. Unexpected status: 200
+		repoName := ctx.Repo.Repository.FullName()
+		url := strings.TrimSuffix(setting.AppURL, "/") + "/api/v1/repos/" + repoName + "/actions/artifacts/" + fmt.Sprintf("%d", art.ID) + "/zip/raw"
+		ctx.Redirect(url, 302)
+		return
+	}
+	// v3 not supported due to not having one unique id
+	ctx.Error(http.StatusNotFound, "artifact not found", fmt.Errorf("artifact not found"))
+}
+
+// DownloadArtifactRaw Downloads a specific artifact for a workflow run directly.
+func DownloadArtifactRaw(ctx *context.APIContext) {
+	// swagger:operation GET /repos/{owner}/{repo}/actions/artifacts/{artifact_id}/zip/raw repository downloadArtifactRaw
+	// ---
+	// summary: Downloads a specific artifact for a workflow run directly
+	// produces:
+	// - application/json
+	// parameters:
+	// - name: owner
+	//   in: path
+	//   description: name of the owner
+	//   type: string
+	//   required: true
+	// - name: repo
+	//   in: path
+	//   description: name of the repository
+	//   type: string
+	//   required: true
+	// - name: artifact_id
+	//   in: path
+	//   description: id of the artifact
+	//   type: string
+	//   required: true
+	// responses:
+	//   "200":
+	//     description: the artifact content
+	//   "400":
+	//     "$ref": "#/responses/error"
+	//   "404":
+	//     "$ref": "#/responses/notFound"
+
+	artifactID := ctx.PathParamInt64("artifact_id")
+
+	art, ok, err := db.GetByID[actions_model.ActionArtifact](ctx, artifactID)
+	if err != nil {
+		ctx.Error(http.StatusInternalServerError, err.Error(), err)
+		return
+	}
+	if !ok {
+		ctx.Error(http.StatusNotFound, "artifact not found", fmt.Errorf("artifact not found"))
+		return
+	}
+
+	// if artifacts status is not uploaded-confirmed, treat it as not found
+	if art.Status != int64(actions_model.ArtifactStatusUploadConfirmed) {
+		ctx.Error(http.StatusNotFound, "artifact not found", fmt.Errorf("artifact not found"))
+		return
+	}
+	ctx.Resp.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename=%s.zip; filename*=UTF-8''%s.zip", url.PathEscape(art.ArtifactName), art.ArtifactName))
+
+	// Artifacts using the v4 backend are stored as a single combined zip file per artifact on the backend
+	// The v4 backend enshures ContentEncoding is set to "application/zip", which is not the case for the old backend
+	if art.ArtifactName+".zip" == art.ArtifactPath && art.ContentEncoding == "application/zip" {
+		art := art
+		if setting.Actions.ArtifactStorage.ServeDirect() {
```

Comment:
IIRC the "ArtifactStorage download" code has duplicated not one time .... maybe it needs a general function to handle download?